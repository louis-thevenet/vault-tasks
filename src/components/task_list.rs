use color_eyre::Result;
use ratatui::{prelude::*, widgets::*};
use tokio::sync::mpsc::UnboundedSender;
use tracing::info;

use super::Component;
use crate::{action::Action, config::Config};

#[derive(Default)]
pub struct TaskList {
    command_tx: Option<UnboundedSender<Action>>,
    config: Config,
}
impl TaskList {
    pub fn new() -> Self {
        Self::default()
    }
}
impl Component for TaskList {
    fn draw(&mut self, frame: &mut Frame, area: Rect) -> Result<()> {
        let surrounding_block = Block::default()
            .borders(Borders::ALL)
            .title("Here is a list of tasks");
        let items = ["Task 1", "Task 2", "Task 3"];
        let list = List::new(items).block(surrounding_block);
        frame.render_widget(list, area);
        Ok(())
    }
}
