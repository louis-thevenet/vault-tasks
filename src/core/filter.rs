use crate::core::TasksConfig;

use super::{
    date::Date,
    parser::task::parse_task,
    task::{State, Task},
    vault_data::VaultData,
};

#[derive(Default, PartialEq, Eq, Debug)]
pub struct Filter {
    pub task: Task,
    state: Option<State>,
}

/// Parses a [`Task`] from an input `&str`. Returns the `Task` and whether the input specify a task state (- [X] or - [ ]) or not.
#[must_use]
pub fn parse_search_input(input: &str, config: &TasksConfig) -> Filter {
    // Are searching for a specific state ?
    let has_state = input.starts_with("- [");

    // Make the input parsable, add a task state if needed
    let input_value = format!("{}{}", if has_state { "" } else { "- [ ]" }, input);

    // Parse the input
    let task = match parse_task(&mut input_value.as_str(), String::new(), config) {
        Ok(t) => t,
        Err(_e) => Task {
            name: String::from("Incomplete search prompt"),
            ..Default::default()
        },
    };
    Filter {
        task: task.clone(),
        state: if has_state { Some(task.state) } else { None },
    }
}

fn filter_task(task: &Task, filter: &Filter) -> bool {
    let state_match = filter.state.is_none()
        || filter.state.clone().is_some_and(|state| {
            // This is not really satisfying as you can't
            // match only incomplete tasks for example
            // I should add a config option for it
            matches!(
                (state, &task.state),
                (
                    State::ToDo | State::Incomplete,
                    State::ToDo | State::Incomplete
                ) | (State::Done | State::Canceled, State::Done | State::Canceled)
            )
        });

    let name_match = names_match(&task.name, &filter.task.name);

    let today_flag_match = if filter.task.is_today {
        task.is_today
    } else {
        true
    };

    let date_match = match (task.due_date.clone(), filter.task.due_date.clone()) {
        (_, None) => true,
        (Some(Date::DayTime(task_date)), Some(Date::DayTime(search_date)))
            if task_date == search_date =>
        {
            true
        }
        (Some(Date::Day(task_date)), Some(Date::Day(search_date))) if task_date == search_date => {
            true
        }
        (_, _) => false,
    };

    let tags_match = filter
        .task
        .tags
        .clone()
        .unwrap_or_default()
        .iter()
        .all(|t| {
            task.tags
                .clone()
                .unwrap_or_default()
                .iter()
                .any(|x| x.to_lowercase().contains(&t.to_lowercase()))
        });

    let priority_match = if filter.task.priority > 0 {
        filter.task.priority == task.priority
    } else {
        true
    };

    state_match && name_match && today_flag_match && date_match && tags_match && priority_match
}

fn names_match(name: &str, filter_name: &str) -> bool {
    if filter_name.is_empty() {
        true
    } else {
        // for each word of the filter_task, if at least one
        // matches in the task, then validate
        filter_name
            .to_lowercase()
            .split_whitespace()
            .filter(|w| name.to_lowercase().contains(w))
            .count()
            > 0
    }
}

/// Collects all tasks matching the provided `Filter` from the `VaultData` in a `Vec<Task>`.
/// If `explore_children` is true, it will also explore subtasks of tasks.
fn filter_to_vec_layer(
    vault_data: &VaultData,
    task_filter: &Filter,
    explore_children: bool,
    res: &mut Vec<Task>,
) {
    match vault_data {
        VaultData::Directory(_, children) | VaultData::Header(_, _, children) => {
            for c in children {
                filter_to_vec_layer(&c.clone(), task_filter, explore_children, res);
            }
        }
        VaultData::Task(task) => {
            if explore_children {
                task.subtasks.iter().for_each(|t| {
                    filter_to_vec_layer(
                        &VaultData::Task(t.clone()),
                        task_filter,
                        explore_children,
                        res,
                    );
                });
            }

            if filter_task(task, task_filter) {
                res.push(task.clone());
            }
        }
        VaultData::Tracker(_tracker) => (), // Don't collect trackers in the result
                                            // It's only used by the Filter and Calendar
                                            // tabs and we don't want to display trackers there
    }
}

/// Will return a `Vec<Task>` matching the given `Filter` from the `VaultData`
pub fn filter_to_vec(vault_data: &VaultData, filter: &Filter) -> Vec<Task> {
    let res = &mut vec![];
    filter_to_vec_layer(vault_data, filter, true, res);
    res.clone()
}

/// Filters a `VaultData` structure based on the provided `Filter`.
/// Only keeps the `VaultData` entries that match the filter criteria.
pub fn filter(vault_data: &VaultData, task_filter: &Filter) -> Option<VaultData> {
    match vault_data {
        VaultData::Header(level, name, children) => {
            let mut actual_children = vec![];
            for child in children {
                let child_clone = child.clone();
                if let Some(child) = filter(&child_clone, task_filter) {
                    actual_children.push(child);
                }
            }
            if actual_children.is_empty() {
                None
            } else {
                Some(VaultData::Header(*level, name.to_string(), actual_children))
            }
        }
        VaultData::Directory(name, children) => {
            let mut actual_children = vec![];
            for child in children {
                let child_clone = child.clone();
                if let Some(child) = filter(&child_clone, task_filter) {
                    actual_children.push(child);
                }
            }
            if actual_children.is_empty() {
                None
            } else {
                Some(VaultData::Directory(name.to_string(), actual_children))
            }
        }
        VaultData::Task(task) => {
            if filter_task(task, task_filter) {
                Some(vault_data.clone())
            } else {
                let mut actual_children = vec![];
                for child in &task.subtasks {
                    if let Some(VaultData::Task(child)) =
                        filter(&VaultData::Task(child.clone()), task_filter)
                    {
                        actual_children.push(child);
                    }
                }
                if actual_children.is_empty() {
                    return None;
                }
                Some(VaultData::Task(Task {
                    subtasks: actual_children,
                    ..task.clone()
                }))
            }
        }
        VaultData::Tracker(tracker) => {
            // We keep the tracker if its name matches the filter task's name
            // But we don't look at the task's state
            // I might want to refactor the Filter to allow parsing a Tracker from
            // the input string later.
            if names_match(&tracker.name, &task_filter.task.name) {
                Some(VaultData::Tracker(tracker.clone()))
            } else {
                None
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use chrono::NaiveDate;

    use crate::core::{
        TasksConfig,
        date::Date,
        filter::{Filter, filter},
        task::{State, Task},
        vault_data::VaultData,
    };

    use super::{filter_to_vec, parse_search_input};

    #[test]
    fn parse_search_input_test() {
        let input = "- [ ] #tag today name p5";
        let config = TasksConfig::default();
        let res = parse_search_input(input, &config);
        let expected = Filter {
            task: Task {
                due_date: Some(Date::Day(chrono::Local::now().date_naive())),
                name: String::from("name"),
                priority: 5,
                state: State::ToDo,
                tags: Some(vec![String::from("tag")]),
                ..Default::default()
            },
            state: Some(State::ToDo),
        };
        assert_eq!(expected, res);
    }

    #[test]
    fn parse_search_input_no_state_test() {
        let input = "#tag today name p5";
        let config = TasksConfig::default();
        let res = parse_search_input(input, &config);
        let expected = Filter {
            task: Task {
                due_date: Some(Date::Day(chrono::Local::now().date_naive())),
                name: String::from("name"),
                priority: 5,
                state: State::ToDo,
                tags: Some(vec![String::from("tag")]),
                ..Default::default()
            },
            state: None,
        };
        assert_eq!(expected, res);
    }

    #[test]
    fn filter_tags_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "test 1".to_string(),
                                    line_number: Some(8),
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "test 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = vec![
            Task {
                name: "test 2".to_string(),
                line_number: Some(8),
                tags: Some(vec!["test".to_string()]),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            },
            Task {
                name: "test 3".to_string(),
                line_number: Some(8),
                tags: Some(vec!["test".to_string()]),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            },
        ];
        let res = filter_to_vec(
            &input,
            &Filter {
                task: Task {
                    name: String::new(),
                    tags: Some(vec!["test".to_string()]),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
    #[test]
    fn filter_names_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "hfdgqskhjfg1".to_string(),
                                    line_number: Some(8),
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "test 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = vec![
            Task {
                name: "test 2".to_string(),
                line_number: Some(8),
                tags: Some(vec!["test".to_string()]),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            },
            Task {
                name: "test 3".to_string(),
                line_number: Some(8),
                tags: Some(vec!["test".to_string()]),
                description: Some("test\ndesc".to_string()),
                ..Default::default()
            },
        ];
        let res = filter_to_vec(
            &input,
            &Filter {
                task: Task {
                    name: String::from("test"),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
    #[test]
    fn filter_due_date_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "hfdgqskhjfg1".to_string(),
                                    line_number: Some(8),
                                    due_date: Some(Date::Day(
                                        NaiveDate::from_ymd_opt(2020, 2, 2).unwrap(),
                                    )),
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "test 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = vec![Task {
            name: "hfdgqskhjfg1".to_string(),
            line_number: Some(8),
            due_date: Some(Date::Day(NaiveDate::from_ymd_opt(2020, 2, 2).unwrap())),
            description: Some("test\ndesc".to_string()),
            ..Default::default()
        }];
        let res = filter_to_vec(
            &input,
            &Filter {
                task: Task {
                    due_date: Some(Date::Day(NaiveDate::from_ymd_opt(2020, 2, 2).unwrap())),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
    #[test]
    fn filter_full_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "real target".to_string(),
                                    line_number: Some(8),
                                    tags: Some(vec!["test".to_string()]),
                                    due_date: Some(Date::Day(
                                        NaiveDate::from_ymd_opt(2020, 2, 2).unwrap(),
                                    )),
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "false target 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = vec![Task {
            name: "real target".to_string(),
            line_number: Some(8),
            due_date: Some(Date::Day(NaiveDate::from_ymd_opt(2020, 2, 2).unwrap())),
            tags: Some(vec!["test".to_string()]),
            description: Some("test\ndesc".to_string()),
            ..Default::default()
        }];
        let res = filter_to_vec(
            &input,
            &Filter {
                task: Task {
                    name: String::from("target"),
                    tags: Some(vec!["test".to_string()]),
                    due_date: Some(Date::Day(NaiveDate::from_ymd_opt(2020, 2, 2).unwrap())),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
    #[test]
    fn filter_subtasks_test() {
        let input = VaultData::Directory(
            "test".to_owned(),
            vec![
                VaultData::Header(
                    0,
                    "Test".to_string(),
                    vec![
                        VaultData::Header(
                            1,
                            "1".to_string(),
                            vec![VaultData::Header(
                                2,
                                "2".to_string(),
                                vec![VaultData::Task(Task {
                                    name: "task".to_string(),
                                    line_number: Some(8),
                                    tags: Some(vec!["test".to_string()]),
                                    description: Some("test\ndesc".to_string()),
                                    subtasks: vec![Task {
                                        name: "subtask".to_string(),
                                        ..Default::default()
                                    }],
                                    ..Default::default()
                                })],
                            )],
                        ),
                        VaultData::Header(
                            1,
                            "1.2".to_string(),
                            vec![
                                VaultData::Header(3, "3".to_string(), vec![]),
                                VaultData::Header(
                                    2,
                                    "4".to_string(),
                                    vec![VaultData::Task(Task {
                                        name: "false target 2".to_string(),
                                        line_number: Some(8),
                                        tags: Some(vec!["test".to_string()]),
                                        description: Some("test\ndesc".to_string()),
                                        ..Default::default()
                                    })],
                                ),
                            ],
                        ),
                    ],
                ),
                VaultData::Task(Task {
                    name: "test 3".to_string(),
                    line_number: Some(8),
                    tags: Some(vec!["test".to_string()]),
                    description: Some("test\ndesc".to_string()),
                    ..Default::default()
                }),
            ],
        );
        let expected = Some(VaultData::Directory(
            "test".to_owned(),
            vec![VaultData::Header(
                0,
                "Test".to_string(),
                vec![VaultData::Header(
                    1,
                    "1".to_string(),
                    vec![VaultData::Header(
                        2,
                        "2".to_string(),
                        vec![VaultData::Task(Task {
                            name: "task".to_string(),
                            line_number: Some(8),
                            tags: Some(vec!["test".to_string()]),
                            description: Some("test\ndesc".to_string()),
                            subtasks: vec![Task {
                                name: "subtask".to_string(),
                                ..Default::default()
                            }],
                            ..Default::default()
                        })],
                    )],
                )],
            )],
        ));
        let res = filter(
            &input,
            &Filter {
                task: Task {
                    name: String::from("subtask"),
                    ..Default::default()
                },
                state: None,
            },
        );
        assert_eq!(res, expected);
    }
}
