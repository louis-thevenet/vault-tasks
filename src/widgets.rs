pub mod task_list;
pub mod task_list_item;
