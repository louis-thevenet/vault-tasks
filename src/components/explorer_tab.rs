use color_eyre::eyre::bail;
use color_eyre::Result;
use crossterm::event::Event;
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders, Paragraph};
use tokio::sync::mpsc::UnboundedSender;
use tracing::{debug, error};

use tui_input::backend::crossterm::EventHandler;
use tui_widget_list::{ListBuilder, ListState, ListView};

use super::Component;

use crate::task_core::filter::parse_search_input;
use crate::task_core::vault_data::VaultData;
use crate::task_core::TaskManager;
use crate::widgets::search_bar::SearchBar;
use crate::widgets::task_list::TaskList;
use crate::{action::Action, config::Config};

#[derive(Default)]
pub struct ExplorerTab<'a> {
    command_tx: Option<UnboundedSender<Action>>,
    config: Config,
    is_focused: bool,
    task_mgr: TaskManager,
    current_path: Vec<String>,
    state_left_view: ListState,
    entries_left_view: Vec<(String, String)>,
    state_center_view: ListState,
    entries_center_view: Vec<(String, String)>,
    entries_right_view: Vec<VaultData>,
    search_bar_widget: SearchBar<'a>,
}

impl<'a> ExplorerTab<'a> {
    pub fn new() -> Self {
        Self::default()
    }

    fn enter_selected_entry(&mut self) -> Result<()> {
        // Update path with selected entry
        self.current_path.push(
            self.entries_center_view[self.state_center_view.selected.unwrap_or_default()]
                .1
                .clone(),
        );

        // Can we enter ?
        if !self.task_mgr.can_enter(&self.current_path) {
            self.current_path.pop();
            debug!("Coudln't enter: {:?}", self.current_path);
            return Ok(());
        }

        // Update selections
        self.state_left_view
            .select(Some(self.state_center_view.selected.unwrap_or_default()));
        self.state_center_view.select(Some(0));

        debug!("Entering: {:?}", self.current_path);

        // Update entries
        self.update_entries()
    }
    fn leave_selected_entry(&mut self) -> Result<()> {
        if self.current_path.is_empty() {
            return Ok(());
        }

        self.current_path.pop().unwrap_or_default();
        // Update index of selected entry to previous selected entry
        self.state_center_view.select(self.state_left_view.selected);

        self.update_entries()?;

        // Find previously selected entry
        if let Some(new_previous_entry) = self.current_path.last() {
            self.state_left_view.select(Some(
                self.entries_left_view
                    .clone()
                    .into_iter()
                    .enumerate()
                    .find(|(_, entry)| &entry.1 == new_previous_entry)
                    .unwrap_or_default()
                    .0,
            ));
        }
        Ok(())
    }

    fn update_entries(&mut self) -> Result<()> {
        debug!("Updating entries");

        if self.current_path.is_empty() {
            // Vault root
            self.entries_left_view = vec![];
        } else {
            self.entries_left_view = self
                .task_mgr
                .get_explorer_entries(&self.current_path[0..self.current_path.len() - 1])?;
        }
        self.entries_center_view =
            if let Ok(res) = self.task_mgr.get_explorer_entries(&self.current_path) {
                res
            } else {
                self.leave_selected_entry()?;
                return Ok(());
            };
        self.update_preview();
        Ok(())
    }

    fn get_preview_path(&self) -> Result<Vec<String>> {
        let mut path_to_preview = self.current_path.clone();
        if self.entries_center_view.is_empty() {
            bail!("Error: No selected entry")
        }
        path_to_preview.push(
            self.entries_center_view[self.state_center_view.selected.unwrap_or_default()]
                .1
                .clone(),
        );
        Ok(path_to_preview)
    }

    fn update_preview(&mut self) {
        debug!("Updating preview");
        let Ok(path_to_preview) = self.get_preview_path() else {
            return;
        };

        self.entries_right_view = self
            .task_mgr
            .get_vault_data_from_path(&path_to_preview)
            .unwrap_or_default();
    }
    fn build_list(
        entries_to_display: Vec<String>,
        surrouding_block: Block<'_>,
    ) -> ListView<'_, Paragraph<'_>> {
        let item_count = entries_to_display.len();

        let builder = ListBuilder::new(move |context| {
            let mut item = Paragraph::new(entries_to_display[context.index].clone());
            if context.is_selected {
                item = item.style(
                    Style::default()
                        .bg(Color::Rgb(255, 153, 0))
                        .fg(Color::Rgb(28, 28, 32)),
                );
            };
            let main_axis_size = 1;
            (item, main_axis_size)
        });

        ListView::new(builder, item_count).block(surrouding_block)
    }
    fn apply_prefixes(entries: &[(String, String)]) -> Vec<String> {
        entries
            .iter()
            .map(|item| format!("{} {}", item.0, item.1))
            .collect()
    }
    pub fn render_footer(area: Rect, frame: &mut Frame) {
        Line::raw("Press hjkl|◄▼▲▶ to move")
            .centered()
            .render(area, frame.buffer_mut());
    }
}

impl<'a> Component for ExplorerTab<'a> {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) -> Result<()> {
        self.command_tx = Some(tx);
        Ok(())
    }

    fn register_config_handler(&mut self, config: Config) -> Result<()> {
        self.task_mgr = TaskManager::load_from_config(&config)?;
        self.config = config;
        self.task_mgr.current_filter = Some(parse_search_input(
            self.search_bar_widget.input.value(),
            &self.config,
        ));
        self.update_entries()?;
        self.state_center_view.selected = Some(0);
        Ok(())
    }

    fn escape_editing_mode(&self) -> Vec<Action> {
        vec![Action::Enter, Action::Escape]
    }
    fn editing_mode(&self) -> bool {
        self.is_focused && self.search_bar_widget.is_focused
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>> {
        if self.is_focused {
            debug!("action in explorer:{action}");
            match action {
                Action::FocusFilter => self.is_focused = false,

                Action::Enter | Action::Escape if self.search_bar_widget.is_focused => {
                    self.search_bar_widget.is_focused = !self.search_bar_widget.is_focused;
                }
                Action::Search => {
                    self.search_bar_widget.is_focused = !self.search_bar_widget.is_focused;
                }
                Action::Key(key_event) if self.search_bar_widget.is_focused => {
                    self.search_bar_widget
                        .input
                        .handle_event(&Event::Key(key_event));

                    // Update search input in TaskManager
                    self.task_mgr.current_filter = Some(parse_search_input(
                        self.search_bar_widget.input.value(),
                        &self.config,
                    ));
                    self.current_path = vec![];
                    self.state_center_view.select(Some(0));
                    self.state_left_view.select(None);
                    self.update_entries()?;
                }

                Action::Up => {
                    self.state_center_view.previous();
                    self.update_preview();
                }
                Action::Down => {
                    self.state_center_view.next();
                    self.update_preview();
                }
                Action::Right | Action::Enter => self.enter_selected_entry()?,
                Action::Left | Action::Escape => self.leave_selected_entry()?,
                Action::Cancel => {
                    self.leave_selected_entry()?;
                }
                Action::Help => todo!(),
                _ => (),
            }
        } else if action == Action::FocusExplorer {
            self.is_focused = true;
        }
        Ok(None)
    }

    fn draw(&mut self, frame: &mut Frame, _area: Rect) -> Result<()> {
        if !self.is_focused {
            return Ok(());
        }
        if self.entries_center_view.is_empty() {
            error!("Center view is empty"); // is it always an error ?
            self.update_entries()?;
            self.state_center_view.selected = Some(0);
        }

        let vertical = Layout::vertical([
            Constraint::Length(1),
            Constraint::Min(0),
            Constraint::Length(1),
            Constraint::Length(1),
        ]);
        let [_header_area, inner_area, footer_area, _tab_footer_area] =
            vertical.areas(frame.area());

        Self::render_footer(footer_area, frame);

        let [search_path_area, explorer_area] =
            Layout::vertical(vec![Constraint::Length(3), Constraint::Percentage(100)])
                .areas(inner_area);

        let [path_area, search_area] =
            Layout::horizontal(vec![Constraint::Percentage(70), Constraint::Percentage(30)])
                .areas(search_path_area);

        // Main Layout
        let [previous_area, current_area, preview_area] = Layout::horizontal(vec![
            Constraint::Percentage(10),
            Constraint::Percentage(30),
            Constraint::Percentage(60),
        ])
        .areas(explorer_area);

        // Search Bar
        if self.search_bar_widget.is_focused {
            let width = search_area.width.max(3) - 3; // 2 for borders, 1 for cursor
            let scroll = self.search_bar_widget.input.visual_scroll(width as usize);

            // Make the cursor visible and ask tui-rs to put it at the specified coordinates after rendering
            frame.set_cursor_position((
                // Put cursor past the end of the input text
                search_area.x.saturating_add(
                    ((self.search_bar_widget.input.visual_cursor()).max(scroll) - scroll) as u16,
                ) + 1,
                // Move one line down, from the border to the input line
                search_area.y + 1,
            ));
        }

        self.search_bar_widget.block = Some(Block::bordered().title("Search").style(
            Style::new().fg(if self.search_bar_widget.is_focused {
                Color::Rgb(255, 153, 0)
            } else {
                Color::default()
            }),
        ));
        self.search_bar_widget
            .render(search_area, frame.buffer_mut());

        // Current path
        frame.render_widget(
            Paragraph::new(format!("\n./{}", self.current_path.join("/"))),
            path_area,
        );

        // Left Block
        let left_entries_list = Self::build_list(
            Self::apply_prefixes(&self.entries_left_view),
            Block::default().borders(Borders::RIGHT),
        );
        let state = &mut self.state_left_view;
        left_entries_list.render(previous_area, frame.buffer_mut(), state);

        // Center Block
        let lateral_entries_list = Self::build_list(
            Self::apply_prefixes(&self.entries_center_view),
            Block::default().borders(Borders::RIGHT),
        );
        let state = &mut self.state_center_view;
        lateral_entries_list.render(current_area, frame.buffer_mut(), state);

        // Right Block

        match self.entries_right_view.first() {
            Some(VaultData::Task(_) | VaultData::Header(_, _, _)) => {
                TaskList::new(&self.config, &self.entries_right_view)
                    .render(preview_area, frame.buffer_mut());
            }
            Some(VaultData::Directory(_, _)) => Self::build_list(
                Self::apply_prefixes(
                    &self
                        .task_mgr
                        .get_explorer_entries(
                            &self
                                .get_preview_path()
                                .unwrap_or_else(|_| self.current_path.clone()),
                        )
                        .unwrap_or_default(),
                ),
                Block::new(),
            )
            .render(preview_area, frame.buffer_mut(), &mut ListState::default()),
            None => (),
        }

        Ok(())
    }
}
