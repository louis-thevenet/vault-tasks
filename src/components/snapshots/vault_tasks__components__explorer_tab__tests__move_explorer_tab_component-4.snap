---
source: src/components/explorer_tab.rs
expression: terminal.backend()
---
"                                                                                                                                                                "
"./chocolate_lava_cake.md/Recipe: Chocolate Lava Cake                                                            ┌Search────────────────────────────────────────┐"
"                                                                                                                │- [ ]                                         │"
"                                                                                                                └──────────────────────────────────────────────┘"
"# Recipe: Choco│❌ Whisk eggs, yolks, sugar, and vanilla:      │┌❌ Bake cakes:────────────────────────────────────────────────────────────────────────────────┐" Hidden by multi-width symbols: [(17, " "), (66, "c")]
"               │❌ Bake cakes:                                 ││📅 2024/10/01                                                                                 │" Hidden by multi-width symbols: [(17, " "), (66, " ")]
"               │❌ Clean up:                                   ││#cooking                                                                                      │" Hidden by multi-width symbols: [(17, " ")]
"               │                                               ││┌✅ Divide the batter evenly among the prepared ramekins.────────────────────────────────────┐│" Hidden by multi-width symbols: [(67, "│")]
"               │                                               │││📅 2024/10/01                                                                               ││" Hidden by multi-width symbols: [(67, "│")]
"               │                                               ││└────────────────────────────────────────────────────────────────────────────────────────────┘│"
"               │                                               ││┌✅ Bake for 12 minutes, or until the edges are set but the centers are still gooey.─────────┐│" Hidden by multi-width symbols: [(67, "│")]
"               │                                               │││📅 2024/10/01                                                                               ││" Hidden by multi-width symbols: [(67, "│")]
"               │                                               ││└────────────────────────────────────────────────────────────────────────────────────────────┘│"
"               │                                               ││┌❌ Allow to cool for about 1 minute before dusting with confectioners' sugar and inverting o┐│" Hidden by multi-width symbols: [(67, "│")]
"               │                                               │││📅 2024/10/01                                                                               ││" Hidden by multi-width symbols: [(67, "│")]
"               │                                               ││└────────────────────────────────────────────────────────────────────────────────────────────┘│"
"               │                                               │└──────────────────────────────────────────────────────────────────────────────────────────────┘"
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"               │                                               │                                                                                                "
"                                                                    Press hjkl|◄▼▲▶ to move                                                                     "
"                                                                                                                                                                "
